File: frontend/src/api/agents.ts
import client from "./client";
import type { Agent, AgentCreate, AgentUpdate } from "@/types/agent";

export async function getAgents(): Promise<Agent[]> {
  const response = await client.get<Agent[]>("/agents");
  return response.data;
}

export async function getAgent(id: string): Promise<Agent> {
  const response = await client.get<Agent>(`/agents/${id}`);
  return response.data;
}

export async function createAgent(data: AgentCreate): Promise<Agent> {
  const response = await client.post<Agent>("/agents", data);
  return response.data;
}

export async function updateAgent(
  id: string,
  data: AgentUpdate
): Promise<Agent> {
  const response = await client.put<Agent>(`/agents/${id}`, data);
  return response.data;
}

export async function deleteAgent(id: string): Promise<void> {
  await client.delete(`/agents/${id}`);
}

File: backend/app/api/agents.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.schemas.agent import AgentCreate, AgentResponse, AgentUpdate
from app.services.agent_service import AgentService

router = APIRouter()


@router.get("/", response_model=list[AgentResponse])
async def list_agents(db: AsyncSession = Depends(get_db)):
    agents = await AgentService.get_all(db)
    return agents


@router.post("/", response_model=AgentResponse, status_code=201)
async def create_agent(data: AgentCreate, db: AsyncSession = Depends(get_db)):
    agent = await AgentService.create(db, data)
    return agent


@router.get("/{agent_id}", response_model=AgentResponse)
async def get_agent(agent_id: str, db: AsyncSession = Depends(get_db)):
    agent = await AgentService.get_by_id(db, agent_id)
    if agent is None:
        raise HTTPException(status_code=404, detail="Agent not found")
    return agent


@router.put("/{agent_id}", response_model=AgentResponse)
async def update_agent(
    agent_id: str,
    data: AgentUpdate,
    db: AsyncSession = Depends(get_db),
):
    agent = await AgentService.update(db, agent_id, data)
    if agent is None:
        raise HTTPException(status_code=404, detail="Agent not found")
    return agent


@router.delete("/{agent_id}", status_code=204)
async def delete_agent(agent_id: str, db: AsyncSession = Depends(get_db)):
    deleted = await AgentService.delete(db, agent_id)
    if not deleted:
        raise HTTPException(status_code=404, detail="Agent not found")
